server/storage/couchdb.go

```diff
@@ -0,0 +1,591 @@
+	Version        int           `json:"version"`
+	SHA256         string        `json:"sha256"`
+	Data           []byte        `json:"data"`
+	TSchecksum     string        `json:"timestamp_checksum"`
+}
+
+// TableName returns the table name for the record type
+func (r CDBTUFFile) TableName() string {
+	return TUFFileTableName
+}
+
+// CChange defines the the fields required for an object in the changefeed
+type CChange struct {
+	couchdb.ID
+	CreatedAt time.Time `json:"created_at"`
+	GUN       string    `json:"gun"`
+	Version   int       `json:"version"`
+	SHA256    string    `json:"sha256"`
+	Category  string    `json:"category"`
+}
+
+// TableName sets a specific table name for Changefeed
+func (cdb CChange) TableName() string {
+	return ChangefeedTableName
+}
+
+// unmarshal in an anonymous struct
+func cdbTUFFileFromJSON(data []byte) (interface{}, error) {
+	a := struct {
+		CreatedAt  time.Time `json:"created_at"`
+		UpdatedAt  time.Time `json:"updated_at"`
+		DeletedAt  time.Time `json:"deleted_at"`
+		Gun        string    `json:"gun"`
+		Role       string    `json:"role"`
+		Version    int       `json:"version"`
+		SHA256     string    `json:"sha256"`
+		Data       []byte    `json:"data"`
+		TSchecksum string    `json:"timestamp_checksum"`
+	}{}
+	if err := json.Unmarshal(data, &a); err != nil {
+		return CDBTUFFile{}, err
+	}
+	return CDBTUFFile{
+		Timing: couchdb.Timing{
+			CreatedAt: a.CreatedAt,
+			UpdatedAt: a.UpdatedAt,
+			DeletedAt: a.DeletedAt,
+		},
+		GunRoleVersion: []interface{}{a.Gun, a.Role, a.Version},
+		Gun:            a.Gun,
+		Role:           a.Role,
+		Version:        a.Version,
+		SHA256:         a.SHA256,
+		Data:           a.Data,
+		TSchecksum:     a.TSchecksum,
+	}, nil
+}
+
+func cdbChangeFromJSON(data []byte) (interface{}, error) {
+	res := CChange{}
+	if err := json.Unmarshal(data, &res); err != nil {
+		return CChange{}, err
+	}
+	return res, nil
+}
+
+// CouchDB implements a MetaStore against the Couchdb Database
+type CouchDB struct {
+	dbName   string
+	client   *kivik.Client
+	user     string
+	password string
+}
+
+// NewCouchDBStorage initializes a CouchDB object
+func NewCouchDBStorage(dbName, user, password string, client *kivik.Client) CouchDB {
+	return CouchDB{
+		dbName:   dbName,
+		client:   client,
+		user:     user,
+		password: password,
+	}
+}
+
+func genID(gun, role string, version int) string {
+	return fmt.Sprintf("%s.%s.%d", gun, role, version)
+}
+
+// UpdateCurrent adds new metadata version for the given GUN if and only
+// if it's a new role, or the version is greater than the current version
+// for the role. Otherwise an error is returned.
+func (cdb CouchDB) UpdateCurrent(gun data.GUN, update MetaUpdate) error {
+
+	file := CDBTUFFile{}
+
+	db, err := couchdb.DB(cdb.client, cdb.dbName, file.TableName())
+	if err != nil {
+		return err
+	}
+
+	id := genID(gun.String(), update.Role.String(), update.Version)
+
+	row, _ := db.Get(context.TODO(), id)
```

Comment:
I'm not familiar with couch, but rethink basically has an insert but error on conflict, which would mean that we can't get a race where we'd get first, see no conflict, and then insert.  Is there a similar mechanism for CouchDB?  Is that what the rev # does?  (e.g. does the update fail if the latest rev is higher than the rev we attempt to update with?)  Otherwise it seems like 2 updates could race here.

Reply:
The _rev field is the revision of a document, which is some sort of hash over its contents. To updated it, one needs to show the current version, which then automatically changes to another value once the update is done. This excludes concurrent updates and races. 

https://writing.jan.io/2013/12/19/understanding-couchdb-conflicts.html

Reply:
Thank you!  I notice a statement at the bottom stating that if writes happen on two separate CouchDB instances in a cluster, there could be a conflict as well - this blog post seems to suggest only writing to one node by using a load balancer.

However, in a mult-notary server scenario, where each node is speaking to a cluster of CouchDB nodes, this may not make sense.

Does CouchDB have read/write quorums?  I am not entirely sure from reading.  It seems like maybe Cloudant does?  (https://danielwertheim.se/how-to-customize-quorum-with-cloudant-using-mycouch/) But it seems like by default CouchDB may not?  (http://grokbase.com/t/couchdb/dev/153rwp8j3x/could-couchdb-2-0-fix-actual-read-quorum).  If it does, I think we should specify read/write quorums.

From http://guide.couchdb.org/draft/conflicts.html it seems like if two incoming writes happen, both documents will be saved on the server but they will be in conflict.  When returning the current version of the document, one of those conflicting documents will be chosen, and the choice will be the same across all the nodes.

My concern is that if two concurrent writes happen:

1. There will be no error, so both writes will seem to succeed, although in practice only one of those writes will be surfaced since there does not seem to be code in this PR to handle conflicting documents.  I'm not sure what that should look like, even.

2. An update consists of multiple metadata files - this means that the current root may not match the current timestamp if there were a conflict writing either the timestamp or the root metadata, or both.  The timestamp from update A could be chosen as the current document by CouchDB, and the root from update B be chosen as the current document by CouchDB.  This may cause issues when pulling if there is a conflict that results in the current root metadata not matching the current timestamp metadata.  The update logic (let's skip the case where we have any cached metadata for now) is as follows:

    1. Download the current root metadata
    1. Download the current timestamp metadata and verify that it was signed by the right key.
    1. Download the rest of the metadata, do validations, etc...

    If the conflict I mentioned above happened (timestamp from A, root from B), then the client would not be able to update any metadata until the conflict(s) were resolved on the server.

In RethinkDB, this is resolved by using read and write quorums.  2 simultaneous writes can't happen - whomever writes the timestamp first wins the race (because if there is a conflict on timestamp, everything else is rolled back).  And getting the current version of anything or getting a particular version number requires a read quorum.

However, without write quorum, a "transaction" of metadata won't be rolled back if there's a timestamp conflict, and that could mess up `GetCurrent` and `GetVersion` on the server.

Reply:
Here's the configuration of a test database on cloudant.com:

curl https://64b40286-4fad-4454-b691-b0f1b0907376-bluemix:${password}@64b40286-4fad-4454-b691-b0f1b0907376-bluemix.cloudant.com/test

```
{"update_seq":"0-g1AAAAHveJzLYWBg4MhgTmFQTUlKzi9KdUhJMjTSS8rVTU7WTStK1DUw1EvOyS9NScwr0ctLLckBKmdKZEiS____f1YiA0ijClyjBQF9SQpAMskeoZUEO5McQHrjybI2AaS1Hqu1hgT05rEASYYGIAXUPp9kqyHaF0C078e03oAo_Qcg-u-Taf0DiHaQ77MAWOqd2w","db_name":"test","sizes":{"file":1468,"external":0,"active":0},"purge_seq":0,"other":{"data_size":0},"doc_del_count":0,"doc_count":0,"disk_size":1468,"disk_format_version":6,"data_size":0,"compact_running":false,"cluster":{"q":8,"n":3,"w":2,"r":2},"instance_start_time":"0"}
```

There are 3 nodes (n=3) and read quorum is '2' ("r":2) and the write quorum is also '2' ("w":2). I didn't have to set this up as part of the database creation but it was configured like this automatically. So presumably they do w = r = round_down(n / 2) + 1, which results in w + r > n for n odd. So consistency is guaranteed for Cloudant.

I have had my challenges trying to set up a CouchDB cluster using containers on different machines...

Reply:
From some experiments I have done now with CouchDB clustering setup it looks like `r` and `w` are automatically adjusted to `round_down(n/2) + 1` when new nodes are added. We should be fine if now every read and write is happening with these adjusted quorum values.

Reply:
Awesome, thank you for checking.

Reply:
The way I understand it is that CouchDB's and Cloudant's internal configuration enforces read and write consistency. The documentation is here: http://docs.couchdb.org/en/2.1.1/api/database/common.html#get--db

However, the frequently used 'find' operation doesn't enforce read consistency but uses a quorum of '1' instead: http://docs.couchdb.org/en/2.1.1/api/database/find.html#post--db-_find

Reply:
@cyli Following this, should we adjust the read quorum value for the 'find' operation to the proper value to ensure consistency? My feeling is if values returned from the 'find' operation are returned to the user while the data is changing, the consistency wouldn't matter as much than in cases where those data are used internally for further updates, deletes, etc.